### PR TITLE
add important linear algebra operations 

### PR DESCRIPTION
**Pull request by marcelluethi**
This PR adds important linear algebra operations, such as qr, cholesky, svd, solve and eigh. 
Furthermore, it puts the implementation of the `LinearAlgebraTensorOps` (such as `norm`, `det`, `inv`) into the `LinearAlgebra` module and makes them just simple wrappers.

### CODE REVIEW

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+import dimwit.tensor.TensorOps.IsFloating
+import dimwit.tensor.TensorOps.IsNumber
+import me.shadaj.scalapy.py
+import me.shadaj.scalapy.py.SeqConverters
+
+/**  Common linear algebra operations.
+  */
+object LinearAlgebra:
+
+  enum VectorNormType:
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala:19]: This is wrong. Norms should be different functions, not a parameter switching a method's behavior.

Additionally, this is wrong as Ord(1) and Ord(2) are identical to L1 and L2!

Let's improve the interface, maybe:

t.norm(p=2)
t.norm(p=Double.PositiveInfinity)

or 

t.pNorm(p=2)

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    case L1
+    case L2
+    case Ord(p: Double)
+    case Inf
+
+  enum MatrixNormType:
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala:25]: See previous comment. Same error here.

Lets think of better interface, here maybe multiple functions instead of parameter changing behavior.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+  /** Computes the QR factorization of the tensor `t`.
+    *
+    * @param t The input tensor to be factorized. It must be a 2D matrix.
+    * @param mode The mode of the QR factorization (Reduced or Complete).
+    * @return A tuple containing two tensors: the orthogonal matrix Q and the upper-triangular matrix R.
+    *
+    * @see [[https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.linalg.qr.html#jax.numpy.linalg.qr JAX documentation]] for more details on the underlying implementation.
+    */
+  def qr[L1: Label, L2: Label, V: IsFloating](t: Tensor2[L1, L2, V], mode: QRMode = QRMode.Reduced): (q: Tensor2[L1, L2, V], r: Tensor2[L1, L2, V]) =
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: I don't know QR, but checking documentation these types look wrong:

Same names imply to me same shapes, but Q and R have different shapes.

Should be Tensor2[L1, K, V], Tensor2[K, L2, V]
for "reduced", where K is maybe a label we define ourselves or an axis the user must provide?

Should be Tensor2[Prime[L1], L1, V], Tensor2[L1, L2, V]
for "complete"

Different return type indicates to me that these should be two different functions. 

FYI: There is also a missing "raw" option (okay by me to not support full interface, yet).

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    case Complete
+
+  /** Computes the determinant of the tensor `t` along the specified axes (L1, L2)
+    *
+    * @param t The input tensor from which to compute the determinant.
+    * @param axis1 The first axis along which to compute the determinant.
+    * @param axis2 The second axis along which to compute the determinant.
+    * @return A new tensor with the determinant computed, where the two specified axes are removed
+    */
+  def det[T <: Tuple: Labels, L1: Label, L2: Label, V: IsFloating](t: Tensor[T, V], axis1: Axis[L1], axis2: Axis[L2])(using
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: A minimal interface design would require a Tensor2 as a determinant. Even a square matrix.

The type should therefore be Tensor2[L1, L2] or maybe even Tensor2[L, Prime[L]]

Should also have integers implementation, not just Floating-point

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  /** Extracts the diagonal along the given two axes (with optional offset),
+    * replacing them by a new 1D axis labeled L1.
+    *
+    * @param t The input tensor from which to extract the diagonal.
+    * @param axis1 The first axis along which to extract the diagonal.
+    * @param axis2 The second axis along which to extract the diagonal.
+    * @param offset The offset of the diagonal from the main diagonal. Positive values indicate diagonals above the main diagonal, while negative values indicate diagonals below it.
+    * @return A new tensor with the diagonal extracted, where the two specified axes are replaced by a new 1D axis labeled L1.
+    */
+  def diagonal[T <: Tuple, L1: Label, L2: Label, V](t: Tensor[T, V], axis1: Axis[L1], axis2: Axis[L2], offset: Int = 0)(using
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: Maybe the same. Should be defined on Tensor2. Instead of axes parameters user must use vmap.

I guess here it does not require to be squared.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    Tensor(Jax.jnp.diagonal(t.jaxValue, offset = offset, axis1 = ev.indices(0), axis2 = ev.indices(1)))
+
+  /** Computes the inverse of the tensor t along the last two axes.
+    * The first axes of the tensors are preserved, while the last
+    * two axes are replaced by their inverses.
+    * The tensor must be square along the last two axes.
+    *
+    * @return a new tensor with the same shape as t, but with the last two axes replaced by their inverses.
+    */
+  def inv[T <: Tuple: Labels, V: IsFloating](t: Tensor[T, V]): Tensor[T, V] = Tensor(Jax.jnp.linalg.inv(t.jaxValue))
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: I guess this is just wrong as it does not require Axes and it not a Tensor2. So current implementation falls back on last two axes (JAX). Should also be based on Tensor2 type. Tensor must be squared so again L1, L2 or L, Prime[L] must be decided.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+      case MatrixNormType.Spectral  => Tensor0(Jax.jnp.linalg.norm(t.jaxValue, ord = 2))
+      case MatrixNormType.One       => Tensor0(Jax.jnp.linalg.norm(t.jaxValue, ord = 1))
+      case MatrixNormType.Inf       => Tensor0(Jax.jnp.linalg.norm(t.jaxValue, ord = Jax.jnp.inf))
+
+  /** Computes the element-wise norm of the tensor `t` l2 norm along the last axis.
+    * @param t The input tensor for which to compute the norm.
+    *
+    * @return A new 0-D tensor containing the computed norm of the input tensor.
+    */
+  def norm[T <: Tuple, V: IsFloating](t: Tensor[T, V]): Tensor0[V] =
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala:101]: This "norms" all the dimensions by squaring all the numbers and summing them up. 

I think I would not wrap/not support this. This seems to be over permissive implementation and not a mathematical / theoretical construct. 

Note a user can always do "t.pow(2).sum"

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+  /** Cholesky factorization.
+    *
+    * @param t The input tensor to be factorized. It must be a symmetric positive-definite matrix.
+    * @param upper If true, the upper-triangular Cholesky factor is returned
+    * @param symmetrizeInput If true, the input matrix is symmetrized before factorization to ensure numerical stability.
+    * @return a triangular matrix representing the cholesky factor
+    *
+    * @see [[https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.linalg.cholesky.html#jax.numpy.linalg.cholesky JAX documentation]] for more details on the underlying implementation.
+    */
+  def cholesky[L1: Label, L2: Label, V: IsFloating](t: Tensor2[L1, L2, V], upper: Boolean = false, symmetrizeInput: Boolean = true): Tensor2[L1, L2, V] =
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: Is keeping the same labels logical? I would have to think about it. Implementation looks fine.

What would be interesting is if we can express NaN in the value type: Float32 | NaN, as cholesky can return NaN type, but maybe lets push that into the future.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+  /** Computes the eigenvalues and eigenvectors of a symmetric matrix `t`.
+    * @param t The input tensor representing a symmetric matrix.
+    * @param upper If true, the upper-triangular part of the matrix is used.
+    * @param symmetrizeInput If true, the input matrix is symmetrized before computation to ensure numerical stability.
+    * @return A tuple containing two tensors: the eigenvalues and the corresponding eigenvectors of the input matrix.
+    *
+    * @see [[https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.linalg.eigh.html#jax.numpy.linalg.eigh JAX documentation]] for more details on the underlying implementation.
+    */
+  def eigh[L1: Label, L2: Label, V: IsFloating](t: Tensor2[L1, L2, V], upper: Boolean = false, symmetrizeInput: Boolean = true)
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: Again t must be squared. So maybe L and Prime[L]? Here maybe more important than before because of eigenvalues being a vector L1 (or L2).

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    *
+    * @param t The input tensor to be decomposed.
+    * @param fullMatrices  If true, compute the full-sized U and Vh matrices; if false, compute the reduced-sized matrices.
+    * @param hermitian If true, the input is a Hermitian matrix.
+    * @return A tuple containing three tensors: the left singular vectors U, the singular values S, and the right singular vectors Vh.
+    *
+    * @see [[https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.linalg.svd.html#jax.numpy.linalg.svd JAX documentation]] for more details on the underlying implementation.
+    */
+  def svd[L1: Label, L2: Label, V: IsFloating](t: Tensor2[L1, L2, V], fullMatrices: Boolean = false, hermitian: Boolean = false)
+      : (U: Tensor2[L1, L2, V], S: Tensor1[L1, V], Vh: Tensor2[L1, L2, V]) =
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: S: Tensor1[K, V] with K being either L1 or L2 based on which is bigger.

Maybe easiest to resolve by an assert statement, requiring L1 >= L2

Types are wrong if fullMatrices is true. U would have to be L1, L1 and V would be L2, L2. Maybe drop fullMatrices support for now?

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+  /** Solves the linear equation Ax = b for x, where A is a square matrix and b is a vector.
+    *
+    * @param a The input tensor representing the square matrix A.
+    * @param b The input tensor representing the vector b.
+    * @return A new tensor containing the solution vector x.
+    *
+    * @see [[https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.linalg.solve.html#jax.numpy.linalg.solve JAX documentation]] for more details on the underlying implementation.
+    */
+  def solve[L1: Label, L2: Label, V: IsFloating](a: Tensor2[L1, L2, V], b: Tensor1[L1, V]): Tensor1[L2, V] =
```
Comment (benikm91) [inline core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: a is squared so only one label type L, makes return value not ambiguous.

Note: Only supports part of solve interface (b can also be matrix in JAX), but okay for me.

File: core/src/main/scala/dimwit/tensor/tensorops/LinearAlgebraOps.scala [hunk 12]
```diff
@@ -22,100 +19,74 @@ object LinearAlgebraOps:
...
-      *
-      * @param axis1 The first axis along which to compute the trace.
-      * @param axis2 The second axis along which to compute the trace.
-      * @param offset The offset of the diagonal from the main diagonal. Positive values indicate diagonals above the main diagonal, while negative values indicate diagonals below it.
+    /** Computes the trace of the tensor.
       *
-      * @return A new tensor with the trace computed, where the two specified axes are removed, and the remaining axes are preserved.
+      * @see [[LinearAlgebra.trace]] for details
       */
     def trace[L1: Label, L2: Label](axis1: Axis[L1], axis2: Axis[L2], offset: Int = 0)(using
```
Comment (benikm91) [inline core/src/main/scala/dimwit/tensor/tensorops/LinearAlgebraOps.scala]: Okay :) I am the sinner :) I implemented it wrong here already. But we should clean it up.

**Review by benikm91 — CHANGES_REQUESTED** [review]
See comments

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 1 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala:19]: The choice of L1 and L2 as important special cases for Ord(1) and Ord(2) was a deliberate choice. These are the two cases that a user needs most often. Ord instead is used to have arbitrary (even fractional) norms, which are less common. `t.norm(VectorNormType.L2)` is more readable than `t.norm(VectorNormType.Ord(2))` 

I don't see why having an enum to select the type of norm is wrong. It is just a design choice and as discoverable and typesafe as `t.normInfinity` or t.normL2`  I have, however,  no strong opinion on this point and could be easily convinced of another design.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 2 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala:25]: see above answer

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 3 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: This is an excellent point. We need to introduce an additional label LBasis, that is not used but denotes the label for the newly introduced basis dimension. I will add that. 

I was aware of the raw option, but did not see the application for us as the moment.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 4 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: The question here is how close we want to be to the original Jax design. For me, enforcing a Rank 2 Tensor would be fine, as long as we do it consistently.  I would, however, not enforce `Tensor2[L,Prime[L]]` as any invertible linear transformation has a determinant. I don't think introducing  a `SquareTensor` type is worth the added complexity at this points. But we can keep it in mind for the future.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 6 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: Yes, it is maybe better to let that be a Tensor2. Again, it is not as general as Jax, but we can always vmap

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 7 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala:101]: I am unsure about this. it is not as easy as `t.pow(2).sum`, at least not if we only have vector and matrix norms as in the current implementation. We would need nested `reduce` implementations to compute something like this. 


Norm is such a fundamental operation, that it seems useful to have a method that works without much ceremony. 

This method is, how the current `norm` in `LinearAlgebraOps`  is implemented.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 8 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: The label seem fine to me, as the cholesky factors map into the same space. 

I would also push more type safety into the future.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 9 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: I think the correct solution is to introduce two more labels that the user must provide: One for the Tensor1 of Eigenvalues and one for the Tensor2 of EigenVectors. The signature could be 
```scala
 def eigh[L1: Label, L2: Label, LEig: Label, LSpace: Label, V: IsFloating](t: Tensor2[L1, L2, V], eigAxis: Axis[LEig], spaceAxis: Axis[LSpace], upper: Boolean = false, symmetrizeInput: Boolean = true)
      : (eigenvalues: Tensor1[LEig, V], eigenvectors: Tensor2[LSpace, LEig, V])
```

Again, I would postpone enforcing a square matrix.

File: core/src/main/scala/dimwit/linalg/LinearAlgebra.scala (hunk 11 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/linalg/LinearAlgebra.scala]: Again, i think square does not mean that they have to have the same label

File: core/src/main/scala/dimwit/tensor/tensorops/LinearAlgebraOps.scala (hunk 12 above)
Comment (marcelluethi) [reply core/src/main/scala/dimwit/tensor/tensorops/LinearAlgebraOps.scala]: Agree, let's make everything 2D

**Review by benikm91 — APPROVED** [review]
The norm issue is still open; however, I suggest merging this as is, since we can fix things later and should not be slowed down by PRs lying around.
